source/adapters/opencl/command_buffer.cpp

```diff
@@ -1,71 +1,127 @@
 //===--------- command_buffer.cpp - OpenCL Adapter ---------------------===//
 //
-// Copyright (C) 2023 Intel Corporation
-//
-// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM
-// Exceptions. See LICENSE.TXT
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
 // SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
 //
-//===----------------------------------------------------------------------===//
+//===-----------------------------------------------------------------===//
 
 #include "command_buffer.hpp"
 #include "common.hpp"
 
-/// Stub implementations of UR experimental feature command-buffers
-
 UR_APIEXPORT ur_result_t UR_APICALL urCommandBufferCreateExp(
-    [[maybe_unused]] ur_context_handle_t hContext,
-    [[maybe_unused]] ur_device_handle_t hDevice,
+    ur_context_handle_t hContext, ur_device_handle_t hDevice,
     [[maybe_unused]] const ur_exp_command_buffer_desc_t *pCommandBufferDesc,
-    [[maybe_unused]] ur_exp_command_buffer_handle_t *phCommandBuffer) {
+    ur_exp_command_buffer_handle_t *phCommandBuffer) {
 
-  cl_adapter::die("Experimental Command-buffer feature is not "
-                  "implemented for OpenCL adapter.");
-  return UR_RESULT_ERROR_UNSUPPORTED_FEATURE;
+  ur_queue_handle_t queue = nullptr;
+  ur_result_t Result = urQueueCreate(hContext, hDevice, nullptr, &queue);
+  if (Result != UR_RESULT_SUCCESS) {
+    return Result;
+  }
+
+  cl_context clContext = cl_adapter::cast<cl_context>(hContext);
+  cl_ext::clCreateCommandBufferKHR_fn F = nullptr;
+  cl_int Res = cl_ext::getExtFuncFromContext<decltype(F)>(
+      clContext, cl_ext::ExtFuncPtrCache->clCreateCommandBufferKHRCache,
+      cl_ext::CreateCommandBufferName, &F);
+
+  if (!F || Res != CL_SUCCESS)
+    return UR_RESULT_ERROR_INVALID_OPERATION;
+
+  auto clCommandBuffer =
+      F(1, cl_adapter::cast<cl_command_queue *>(&queue), nullptr, &Res);
```

Comment:
Should we be checking the return value from  `Res` here?

Reply:
Fixed, thanks. Added a CL_RETURN_ON_FAILURE_AND_SET_NULL for `Res`.